File: middleware/localization_budget.go
// Package middleware: LocalizationBudget (Sweep TTTT-I) attaches a
// per-request localization counter to ctx + stamps the totals as
// response headers when the handler returns. Wire-up: register on
// every service group that exposes localized payloads (the standard
// `/api/v3/*` + `/internal/*` chains).
//
// Visible headers:
//   X-Kielo-Loc-Refs       — total ref count resolved
//   X-Kielo-Loc-Overrides  — override-store lookups (composite SQL count)
//   X-Kielo-Loc-CacheGets  — Redis MGET / GET round-trips
//   X-Kielo-Loc-Providers  — LLM / opus-mt provider calls
//
// Pre-TTTT-B the GET /concept-hubs/by-concept endpoint emitted
// `X-Kielo-Loc-Overrides: 264` (one per scenario); post-TTTT-B it's
// `X-Kielo-Loc-Overrides: 1`. Dashboards diff these over time to
// detect new N+1 regressions.

package middleware

import (
	"context"
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"

	"github.com/team-kielo-app/kielo-shared/localization"
)

const (
	HeaderKieloLocRefs      = "X-Kielo-Loc-Refs"
	HeaderKieloLocOverrides = "X-Kielo-Loc-Overrides"
	HeaderKieloLocCacheGets = "X-Kielo-Loc-Cachegets"
	HeaderKieloLocProviders = "X-Kielo-Loc-Providers"
)

// LocalizationBudget returns Echo middleware that:
//  1. Attaches a fresh localization budget counter to the request ctx
//     via localization.WithBudget.
//  2. After the handler completes, stamps the totals as response headers.
//
// No-op for handlers that don't invoke the seam — the counter just
// remains at zero. Safe to register globally.
func LocalizationBudget() echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			req := c.Request()
			ctx := localization.WithBudget(req.Context())
			c.SetRequest(req.WithContext(ctx))
			err := next(c)
			stampBudgetHeaders(c.Response().Header(), ctx)
			return err
		}
	}
}

// LocalizationBudgetStdlib is the chi/net-http variant of
// LocalizationBudget for services that don't use Echo (kielo-convo's
// go_orchestrator uses chi). Same contract: stamps the budget
// snapshot as response headers after the wrapped handler returns.
//
// Header stamping is best-effort: chi handlers that call WriteHeader
// before returning will skip the post-handler stamp; the wrapper
// emits the headers eagerly via budgetResponseWriter so they land
// in either flow.
func LocalizationBudgetStdlib(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx := localization.WithBudget(r.Context())
		r = r.WithContext(ctx)
		bw := &budgetResponseWriter{ResponseWriter: w, ctx: ctx}
		next.ServeHTTP(bw, r)
		// Final flush in case the handler never wrote headers
		// (e.g. 204 No Content path with no body).
		bw.flush()
	})
}

// stampBudgetHeaders writes the four loc-budget headers onto h.
// Idempotent — safe to call from both the Echo and stdlib middleware.
func stampBudgetHeaders(h http.Header, ctx context.Context) {
	snap := localization.BudgetSnapshotFromContext(ctx)
	h.Set(HeaderKieloLocRefs, strconv.Itoa(snap.RefsResolved))
	h.Set(HeaderKieloLocOverrides, strconv.Itoa(snap.OverrideLookups))
	h.Set(HeaderKieloLocCacheGets, strconv.Itoa(snap.CacheGets))
	h.Set(HeaderKieloLocProviders, strconv.Itoa(snap.ProviderCalls))
}

// budgetResponseWriter intercepts WriteHeader so we stamp the budget
// just before the status line goes out. Mirrors the standard chi
// pattern of wrapping w to hook the status-emission point.
type budgetResponseWriter struct {
	http.ResponseWriter
	ctx     context.Context
	stamped bool
}

func (b *budgetResponseWriter) WriteHeader(code int) {
	b.flush()
	b.ResponseWriter.WriteHeader(code)
}

func (b *budgetResponseWriter) Write(p []byte) (int, error) {
	b.flush()
	return b.ResponseWriter.Write(p)
}

func (b *budgetResponseWriter) flush() {
	if b.stamped {
		return
	}
	b.stamped = true
	stampBudgetHeaders(b.ResponseWriter.Header(), b.ctx)
}

File: kielo_shared/middleware/localization_budget.py
"""Sweep YYYY: Starlette middleware that stamps per-request
localization budget headers on every response.

Python sibling of the Go ``kielo-shared/middleware/localization_budget.go``
LocalizationBudget Echo middleware + LocalizationBudgetStdlib chi
variant. Reads counters populated by the seam (and the engine's
batch helpers) via ``record_budget``; emits the same
``X-Kielo-Loc-Refs / Overrides / CacheGets / Providers`` header
quartet so observability dashboards aggregate across language
boundaries (Go services + Python engine) without per-service
shape divergence.

Wire-up:

    from kielo_shared.middleware.localization_budget import (
        LocalizationBudgetMiddleware,
    )

    app = FastAPI()
    app.add_middleware(LocalizationBudgetMiddleware)

Place AFTER any auth / language middlewares — the counters are
opened on every request including unauth'd 401 responses (a
malicious client triggering N seam calls before failing auth still
shows up in the metrics).
"""

from __future__ import annotations

from typing import Awaitable, Callable

from kielo_shared.localization.budget import (
    budget_snapshot,
    reset_budget,
    with_budget,
)
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response


HEADER_REFS = "X-Kielo-Loc-Refs"
HEADER_OVERRIDES = "X-Kielo-Loc-Overrides"
HEADER_CACHE_GETS = "X-Kielo-Loc-CacheGets"
HEADER_PROVIDERS = "X-Kielo-Loc-Providers"


class LocalizationBudgetMiddleware(BaseHTTPMiddleware):
    """Per-request localization budget headers."""

    async def dispatch(
        self,
        request: Request,
        call_next: Callable[[Request], Awaitable[Response]],
    ) -> Response:
        token = with_budget()
        try:
            response = await call_next(request)
        finally:
            snapshot = budget_snapshot()
            reset_budget(token)
        # Stamp headers AFTER restoring the contextvar so the snapshot
        # reflects the just-completed handler's work, not any cleanup
        # that may run post-response.
        response.headers[HEADER_REFS] = str(snapshot.refs_resolved)
        response.headers[HEADER_OVERRIDES] = str(snapshot.override_lookups)
        response.headers[HEADER_CACHE_GETS] = str(snapshot.cache_gets)
        response.headers[HEADER_PROVIDERS] = str(snapshot.provider_calls)
        return response
